framework/include/interfaces/TaggingInterface.h

```diff
@@ -173,3 +377,152 @@ class TaggingInterface
+                                          const Indices & dof_indices,
+                                          Real scaling_factor)
+{
+  addResiduals(assembly, residuals, dof_indices, scaling_factor);
+  addJacobian(assembly, residuals, dof_indices, scaling_factor);
+}
+
+template <typename Residuals, typename Indices>
+void
+TaggingInterface::addJacobian(Assembly & assembly,
+                              const Residuals & residuals,
+                              const Indices & dof_indices,
+                              Real scaling_factor)
+{
+  assembly.cacheJacobian(
+      residuals, dof_indices, scaling_factor, Assembly::LocalDataKey{}, _matrix_tags);
+}
+
+template <typename Residuals, typename Indices>
+void
+TaggingInterface::addResidualsAndJacobianWithoutConstraints(Assembly & assembly,
+                                                            const Residuals & residuals,
+                                                            const Indices & dof_indices,
+                                                            Real scaling_factor)
+{
+  addResidualsWithoutConstraints(assembly, residuals, dof_indices, scaling_factor);
+  addJacobianWithoutConstraints(assembly, residuals, dof_indices, scaling_factor);
+}
+
+template <typename Residuals, typename Indices>
+void
+TaggingInterface::addJacobianWithoutConstraints(Assembly & assembly,
+                                                const Residuals & residuals,
+                                                const Indices & dof_indices,
+                                                Real scaling_factor)
+{
+  assembly.cacheJacobianWithoutConstraints(
+      residuals, dof_indices, scaling_factor, Assembly::LocalDataKey{}, _matrix_tags);
+}
+
+inline void
+TaggingInterface::addJacobianElement(Assembly & assembly,
+                                     const Real value,
+                                     const dof_id_type row_index,
+                                     const dof_id_type column_index,
+                                     const Real scaling_factor)
+{
+  assembly.cacheJacobian(
+      row_index, column_index, value * scaling_factor, Assembly::LocalDataKey{}, _matrix_tags);
+}
+
+inline void
+TaggingInterface::addJacobian(Assembly & assembly,
+                              DenseMatrix<Real> & local_k,
+                              const std::vector<dof_id_type> & row_indices,
+                              const std::vector<dof_id_type> & column_indices,
+                              const Real scaling_factor)
+{
+  for (const auto matrix_tag : _matrix_tags)
+    assembly.cacheJacobianBlock(
+        local_k, row_indices, column_indices, scaling_factor, Assembly::LocalDataKey{}, matrix_tag);
+}
+
+template <typename T>
+void
+TaggingInterface::setResidual(SystemBase & sys, const T & residual, MooseVariableFE<T> & var)
+{
+  for (const auto tag_id : _vector_tags)
+    if (sys.hasVector(tag_id))
```

Comment:
I know this is old code; but since this is on your mind... there's no way we can get around this, right? Because we're lazy and we don't do the checking elsewhere?

Reply:
I feel like it's very difficult to know whether the passed in system (developer choice) will have all of the vector tags which a user may have specified in their input file (user choice). So I think this should either stay as-is or potentially be made into an exception/error

Reply:
That's fine; whatever you find appropriate. Just wanted to open the discussion while this is fresh on your mind.

Reply:
Yea it seems like we already have some things which would fail with erroring
```
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: Assertion `sys.hasVector(tag_id)' failed
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: We should not have tags in _vector_tags that do not exist in the system
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: at /home/lindad/projects/ad_kernel_abs_value_ref_23738/framework/build/header_symlinks/TaggingInterface.h, line 541
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: Stack frames: 16
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 0: libMesh::print_trace(std::ostream&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 1: void TaggingInterface::setResidual<double>(SystemBase&, double const&, MooseVariableFE<double>&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 2: NodalBC::computeResidual()
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 3: NonlinearSystemBase::computeNodalBCs(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 4: NonlinearSystemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 5: FEProblemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 6: PODFullSolveMultiApp::computeResidual()
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 7: PODFullSolveMultiApp::solveStep(double, double, bool)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 8: FEProblemBase::execMultiApps(MooseEnumItem, bool)
```